pkg/test/runner/runner.go

```diff
@@ -268,6 +269,9 @@ func (r *Runner) compareResult(exitErr error, tmpPkgPath, resultsPath string) er
 	if exitCode != r.testCase.Config.ExitCode {
 		return fmt.Errorf("actual exit code %d doesn't match expected %d", exitCode, r.testCase.Config.ExitCode)
 	}
+	if exitCode != 0 {
+		return r.compareOutput(stdout, stderr)
+	}
```

Comment:
Even the exit code is not 0, we still need to checkout results and diff. We have structured output in the failure. We also need to make sure the resources are not modified in failure. So no matter what the exit code is, we always need to check everything.